### DISCUSSION TITLE
`$1` necessity in `commit-msg` hook

### DISCUSSION THREAD

**Discussion by privatenumber** [typicode/husky#1108] [Q&A] [answered]
[In the docs](https://typicode.github.io/husky/#:~:text=npx%20%2D%2Dno%2Dinstall%20commitlint%20%2D%2Dedit%20%22%241%22), using `$1` is recommended for the `commit-msg` hook. I'm curious if passing in `$1` is actually necessary here:
```
npx --no-install commitlint --edit $1
```

Upon testing, `$1` always seems to evaluate to `.git/COMMIT_EDITMSG` (which is also the [default value for `commitlint -e`](https://github.com/conventional-changelog/commitlint/blob/c7172022097b11f46b33617e4a94d751243c1049/%40commitlint/cli/src/cli.ts#L52)). Although I have read sources that say `$1` evaluates to the commit message itself (rather than a file path containing it), I have not found that to be the case, which I confirmed with the [Git hooks docs](https://git-scm.com/docs/githooks#_commit_msg):

> It takes a single parameter, the name of the file that holds the proposed commit log message.

Is there an edge case where `$1` evaluates to a different path that this is trying to handle? Or can it be safely omitted?

**Comment by privatenumber** [comment]
@typicode

Pinging you in case you missed this.

**Reply by typicode** [reply]
Hi @privatenumber,

Sorry I don't know. Maybe if you're using Git submodule, path may be different.

In any case, I think I'd stay safe and follow Git hooks docs to avoid any edge case.

**Comment by privatenumber** [comment, marked as answer]
I investigated in the [git source code](https://github.com/git/git) and it seems like `$1` always ends with `/COMMIT_EDITMSG`.

The `.git` directory can be [configured](https://git-scm.com/docs/git-init#Documentation/git-init.txt---separate-git-dirltgit-dirgt) so the directory path may be different. I haven't worked with submodules enough but perhaps this is can be different for submodules too, as you speculate.

Anyway, it should be safe to assume `.git/COMMIT_EDITMSG` as long as you're not changing the `.git` directory.

### Investigation

1. The `commit-msg` commit hook is called here with argument `git_path_commit_editmsg()`:
https://github.com/git/git/blob/715d08a9e51251ad8290b181b6ac3b9e1f9719d7/builtin/commit.c#L1088

2. `git_path_commit_editmsg` is defined using `GIT_PATH_FUNC` here:
https://github.com/git/git/blob/715d08a9e51251ad8290b181b6ac3b9e1f9719d7/sequencer.c#L44

3. `GIT_PATH_FUNC` is defined here:
https://github.com/git/git/blob/715d08a9e51251ad8290b181b6ac3b9e1f9719d7/path.h#L155-L162

4. Since `COMMIT_EDITMSG` is hard-coded to `git_path_commit_editmsg()`, it seems `$1` will always be `$GIT_DIR/COMMIT_EDITMSG` where `$GIT_DIR` [can be configured](https://stackoverflow.com/questions/40561234/can-you-change-git-folder-location).

**Reply by typicode** [reply]
Thanks for the very detailed research :+1:
